### DISCUSSION TITLE
import as node module fails

### DISCUSSION THREAD

**Discussion by geraldo** [maplibre/maplibre-gl-js#1668] [Q&A] [answered]
I try to build the [most basic maplibre example](https://maplibre.org/maplibre-gl-js-docs/example/simple-map/) as node module, but without success. What am I doing wrong?

```
import * as maplibregl from "maplibre-gl";

var map = new maplibregl.Map({
    container: 'map', // container id
    style: 'https://demotiles.maplibre.org/style.json', // style URL
    center: [0, 0], // starting position [lng, lat]
    zoom: 1 // starting zoom
});
```

This throws this error:

`TypeError: maplibregl.Map is not a constructor`

Or is this not working yet, because related to https://github.com/maplibre/maplibre-gl-js/issues/1595?

**Comment by HarelM** [comment]
Have you tried `import maplibre from "maplibre-gl"`?
It probably depends on the packaging system and all kind of other javascript toolings...
You can take a look at the examples to see that this works as expected in plain javascript + html...

**Reply by geraldo** [reply]
> Have you tried `import maplibre from "maplibre-gl"`? It probably depends on the packaging system and all kind of other javascript toolings... You can take a look at the examples to see that this works as expected in plain javascript + html...

Ok, now the error message changed to:

`ReferenceError: devicePixelRatio is not defined`

I know that it works as expected in plain javascript + html, I regulary use it like that. But I'm curious how to make it work with node.

Here my _package.json_:

```
{
  "name": "maplibre-viewer",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "start": "node index.js"
  },
  "author": "",
  "license": "",
  "dependencies": {
    "maplibre-gl": "^2.4.0"
  }
}
```

And my _index.html_:

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Maplibre Viewer</title>
  </head>
  <body>
    <div id="map"></div>
    <script src="./index.js"></script>
  </body>
</html>
```

**Comment by HarelM** [comment]
Since I don't think this is a real issue I'm converting it to a discussion.

**Comment by HarelM** [comment, marked as answer]
maplibre-gl can't run as is on node environment as it uses a lot of things that are available only in the browser.
Having said that, we are able to mock some of the things when we run the jest unit tests and the render-tests which run on node.
Feel free to look at the code there and understand what is needed in order to run it on node environment.
Note that it's not an easy task :-)

**Comment by acalcutt** [comment]
Are you able to use maplibre-native or is there a reason you want maplibre-gl-js ? we just released our first node prerelease there.

**Reply by acalcutt** [reply]
For example, for testing I had made this small test application

https://github.com/acalcutt/maplibre-label-clipping-test/blob/main/index.js

Also, my version of tileserver-gl is using it as an import.
https://github.com/acalcutt/tileserver-gl/blob/main/src/serve_rendered.js

this should work with @maplibre/maplibre-gl-native now that we have released it
